### PR TITLE
services/horizon: Refactors integration test framework to allow custom parameter testing.

### PR DESCRIPTION
**Pull request by Shaptic**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What
This introduces the `IntegrationTest.Config.HorizonParameters` field, which allows you to customize the parameters passed to Horizon, and refactors the tests to use these accordingly. It also introduces `FatalTestSuite`, which is a way to run integration tests and see if they *should* fail.

**Note**: This is mostly intended a design proposal and an opportunity to get feedback before I dive into actual testing of the combinatorial explosion of parameters we have. There are a few tests and validators included to demonstrate how said tests would look.

### Why
In order to facilitate configuration testing (see #3193), we need to be able to run Horizon with arbitrary combinations of parameters.

### CODE REVIEW

File: ingest/ledgerbackend/toml.go [hunk 1]
```diff
@@ -21,6 +21,10 @@ const (
 	// if LOG_FILE_PATH is omitted stellar core actually defaults to "stellar-core.log"
 	// however, we are overriding this default for captive core
 	defaultLogFilePath = "" // by default we disable logging to a file
+
+	// if DISABLE_XDR_FSYNC is omitted stellar core actually defaults to false
+	// however, we are overriding this default for captive core
+	defaultDisableXDRFsync = true
```
Comment (Shaptic) [inline ingest/ledgerbackend/toml.go:27]: Ignore changes to this file: this is an out-dated merge from master, I think.

File: services/horizon/internal/test/integration/integration.go [hunk 2]
```diff
@@ -189,76 +243,93 @@ func (i *Test) startHorizon(
...
+	var env strings.Builder
+	for key, value := range i.config.HorizonEnvironment {
+		env.WriteString(fmt.Sprintf("%s=%s", key, value))
+	}
+	i.t.Logf("Horizon environmental variables: %s\n", env.String())
+
+	// prepare env
 	cmd.SetArgs(args)
+	for key, value := range i.config.HorizonEnvironment {
+		os.Setenv(key, value)
```
Comment (tamirms) [inline services/horizon/internal/test/integration/integration.go:310]: I think we need to restore the environment variables to what they were after stopping horizon

File: services/horizon/internal/test/integration/integration.go [hunk 3]
```diff
@@ -70,64 +99,92 @@ func NewTestForRemoteHorizon(t *testing.T, horizonURL string, passPhrase string,
...
-	// Only run Stellar Core container and its dependencies
-	runComposeCommand("up", "--detach", "--quiet-pull", "--no-color", "core")
+	if value, ok := config.HorizonParameters[horizon.StellarCoreBinaryPathName]; ok {
+		i.coreConfig.binaryPath = value
+	}
+	if value, ok := config.HorizonParameters[horizon.CaptiveCoreConfigPathName]; ok {
+		i.coreConfig.configPath = value
+	}
 
+	if !bothOrNeitherString(i.coreConfig.binaryPath, i.coreConfig.configPath) {
```
Comment (tamirms) [inline services/horizon/internal/test/integration/integration.go:132]: doesn't horizon enforce this when it parses the command line flags? If so maybe we can remove this code because there's no need to do the check twice

File: services/horizon/internal/test/integration/integration.go [hunk 4]
```diff
@@ -70,64 +99,92 @@ func NewTestForRemoteHorizon(t *testing.T, horizonURL string, passPhrase string,
...
+		t:           t,
+		config:      config,
+		composePath: composePath,
+		passPhrase:  StandaloneNetworkPassphrase,
 	}
 
-	var captiveCoreBinaryPath, captiveCoreConfigPath string
+	// We either test Captive Core through environment variables, or through
+	// custom Horizon parameters.
 	if os.Getenv("HORIZON_INTEGRATION_ENABLE_CAPTIVE_CORE") != "" {
```
Comment (tamirms) [inline services/horizon/internal/test/integration/integration.go:120]: now that we're using this check in multiple places, I think it would make sense to create a global variable:

```golang
var RunWithCaptiveCore = os.Getenv("HORIZON_INTEGRATION_ENABLE_CAPTIVE_CORE") != ""
```

File: services/horizon/internal/integration/parameters_test.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+		horizon.StellarCoreDBURLFlagName:  "",
+	})
+	defer os.RemoveAll(STORAGE_PATH)
+
+	err := test.StartHorizon()
+	assert.NoError(t, err)
+	// FIXME: IntegrationTest needs a big refactor so we can properly wait for
+	//        Captive Core. This needs a lot more thought so this workaround is
+	//        an unfortunate temporary necessity...
+	time.Sleep(10 * time.Second)
```
Comment (tamirms) [inline services/horizon/internal/integration/parameters_test.go:115]: why is this necessary, doesn't `waitForHorizon()` which is called by `StartHorizon()` ensure that horizon is ingesting before it returns?

File: services/horizon/internal/integration/parameters_test.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	// This is a bit of a hacky workaround.
+	//
+	// In CI, we run our integration tests twice: once with Captive Core
+	// enabled, and once without. *These* tests only run with Captive Core
+	// configured properly (specifically, w/ the CAPTIVE_CORE_BIN envvar set).
+	if _, ok := os.LookupEnv("HORIZON_INTEGRATION_ENABLE_CAPTIVE_CORE"); !ok {
+		suite.T().Skip()
+	}
+
+	defer createCaptiveCoreConfig(
```
Comment (tamirms) [inline services/horizon/internal/integration/parameters_test.go:76]: this is a bit confusing because when I read `defer createCaptiveCoreConfig....` I assume that `createCaptiveCoreConfig()` is the operation that is getting deferred. I think it would be cleaner to make `createCaptiveCoreConfig()` write the file and not return anything. Then in the next line you could call `defer os.Remove("./captive-core.toml")`

File: services/horizon/internal/integration/parameters_test.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+		"./captive-core.toml", BUCKET_DIR_DISALLOWED_TOML)()
+
+	const STORAGE_PATH string = "./test_no-bucket-dir"
+	test := NewParameterTest(suite.T(), map[string]string{
+		"captive-core-storage-path":       STORAGE_PATH,
+		horizon.CaptiveCoreConfigPathName: "./captive-core.toml",
+		horizon.StellarCoreBinaryPathName: os.Getenv("CAPTIVE_CORE_BIN"),
+	})
+	defer os.RemoveAll(STORAGE_PATH)
```
Comment (tamirms) [inline services/horizon/internal/integration/parameters_test.go:85]: does "./test_no-bucket-dir" exist? if not then calling `os.RemoveAll(STORAGE_PATH)` is not necessary

File: services/horizon/internal/integration/parameters_test.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+	suite.Exits(func() { test.StartHorizon() })
```
Comment (tamirms) [inline services/horizon/internal/integration/parameters_test.go:87]: I don't understand what's happening here. Could you explain what suite.Exits is doing and why it's necessary to implement the control flow in this way?

File: services/horizon/internal/integration/parameters_test.go (hunk 6 above)
Comment (paulbellamy) [reply services/horizon/internal/integration/parameters_test.go:76]: Or at least:
```go
cleanup := createCaptiveCoreConfig("./captive-core.toml", BUCKET_DIR_DISALLOWED_TOML)
defer cleanup()
```

File: services/horizon/internal/integration/parameters_test.go (hunk 8 above)
Comment (paulbellamy) [reply services/horizon/internal/integration/parameters_test.go:87]: As I understand it, (correct me if I'm wrong @Shaptic) we need this because our config is injected via global variables, and ad-hoc places in the code fetching environment variables. So to run the app with different configs we have to launch a whole new subprocess...

If we had a more controlled way of passing config into horizon we could do something like:
```go
import "github.com/services/horizon/cmd"

c := cmd.RootCmd
c.SetEnv(map[string]string{"ENV_VARS_HERE": "1"}) // Not sure how to do this with cobra
c.SetArgs([]string{"db", "migrate", "up"})
err := c.Execute()
assert.NoError(t, c.Execute()) // or assert.Error(t, c.Execute(), "whatever we expect")
```

But yeah the control flow around the subprocess thing in `Exits` is quite weird either way.

File: services/horizon/internal/integration/parameters_test.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	assert.NoError(t, err)
+	// FIXME: IntegrationTest needs a big refactor so we can properly wait for
+	//        Captive Core. This needs a lot more thought so this workaround is
+	//        an unfortunate temporary necessity...
+	time.Sleep(10 * time.Second)
+
+	runParameterMatrix(test, []ValidatorFunc{
+		func() { validateCaptiveCoreDiskState(test, STORAGE_PATH) },
+		func() { validateNoBucketDirPath(test, STORAGE_PATH) },
+	})
```
Comment (paulbellamy) [inline services/horizon/internal/integration/parameters_test.go:120]: I'd do these as two separate tests, and extract the setup (which is the common part) instead of extracting the test (which is the unique part).

```go

// Should this be a tempdir?
const storagePath string = "./test_captive-core-works"

var defaultCaptiveCoreConfig = map[string]string{
  "captive-core-storage-path":       storagePath,
  "captive-core-reuse-storage-path": "true",
  horizon.StellarCoreBinaryPathName: os.Getenv("CAPTIVE_CORE_BIN"),
  horizon.CaptiveCoreConfigPathName: "./captive-core.toml",
  horizon.StellarCoreURLFlagName:    "",
  horizon.StellarCoreDBURLFlagName:  "",
}

func TestCaptiveCoreCreatesExpectedDirsOnLaunch(t *testing.T) {
  // ensure that running Captive Core creates a
  // sensible directory structure.
  //
  // Pass "rootDirectory" set to whatever it is you pass to
  // "--captive-core-storage-path".
    
  if _, ok := os.LookupEnv("HORIZON_INTEGRATION_ENABLE_CAPTIVE_CORE"); !ok {
    t.Skip() // explained above
  }
  
  cleanup := launchHorizon(t, defaultConfig)
  defer cleanup()

  storageDir := path.Join(rootDir, "captive-core")
  coreConf := path.Join(storageDir, "stellar-core.conf")
  
  assert.DirExists(t, rootDir)
  assert.DirExists(t, storageDir)
  assert.FileExists(t, coreConf)
}

func TestCaptiveCoreConfigExcludesBucketDirPathEntry(t *testing.T) {
  // ensure the Captive Core auto-generated configuration
  // file doesn't contain the BUCKET_DIR_PATH entry, which is forbidden when using
  // Captive Core.
  //
  // Pass "rootDirectory" set to whatever it is you pass to
  // "--captive-core-storage-path".
  
  if _, ok := os.LookupEnv("HORIZON_INTEGRATION_ENABLE_CAPTIVE_CORE"); !ok {
    t.Skip() // explained above
  }
  
  cleanup := launchHorizon(t, defaultConfig)
  defer cleanup()
  
  coreConf := path.Join(rootDir, "captive-core", "stellar-core.conf")
  assert.FileExists(t, coreConf)
  
  result, err := ioutil.ReadFile(coreConf)
  assert.NoError(t, err)
  
  isBucketPathSet := strings.Contains(string(result), "BUCKET_DIR_PATH")
  assert.False(t, isBucketPathSet)
}

// Paul: Or whatever you need to do to launch horizon here.
func launchHorizon(t *testing.T, config map[string]string) (func()) {
  // Paul: Should this be a temp file? Actually, why does this need to be a file?
  cleanupConfig := createCaptiveCoreConfig("./captive-core.toml", CAPTIVE_CORE_CONFIG_STATE_TOML)
  
  test := NewParameterTest(t, config)
  defer os.RemoveAll(storagePath)
  
  err := test.StartHorizon()
  assert.NoError(t, err)
  // FIXME: IntegrationTest needs a big refactor so we can properly wait for
  //        Captive Core. This needs a lot more thought so this workaround is
  //        an unfortunate temporary necessity...
  time.Sleep(10 * time.Second)
  
  // Paul: Do we need to kill horizon when the test finishes?
  return cleanupConfig
}
```

File: services/horizon/internal/test/integration/integration.go [hunk 10]
```diff
@@ -42,14 +44,41 @@ type Config struct {
 	ProtocolVersion       int32
 	SkipContainerCreation bool
 	CoreDockerImage       string
+
+	// Weird naming here because bools default to false, but we want to start
+	// Horizon by default.
+	DontStartHorizon bool
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:50]: Nit/Suggestion:
```suggestion
	SkipHorizon bool
```
Above we called it `SkipContainerCreation`. Maybe there's a similar phrasing here? `SkipHorizonStart`?

File: services/horizon/internal/integration/parameters_test.go [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+	storageDir := path.Join(rootDir, "captive-core")
+	coreConf := path.Join(storageDir, "stellar-core.conf")
+
+	tt.DirExists(rootDir)
+	tt.DirExists(storageDir)
+	tt.FileExists(coreConf)
+}
+
+func createCaptiveCoreConfig(filename, contents string) func() {
+	tomlFile, err := os.Create(filename)
```
Comment (paulbellamy) [inline services/horizon/internal/integration/parameters_test.go:193]: again could this be a tempdir or tempfile to avoid hardcoding paths into tests?

File: services/horizon/internal/test/integration/integration.go [hunk 12]
```diff
@@ -42,14 +44,41 @@ type Config struct {
...
+
+	composePath string
+
 	config        Config
+	coreConfig    CaptiveConfig
 	horizonConfig horizon.Config
-	hclient       *sdk.Client
-	cclient       *stellarcore.Client
+
+	hclient *sdk.Client
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:79]: To match the above `horizonConfig`:
```suggestion
	horizonClient *sdk.Client
```

File: services/horizon/internal/test/integration/integration.go [hunk 13]
```diff
@@ -42,14 +44,41 @@ type Config struct {
...
+	cclient *stellarcore.Client
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:80]:
```suggestion
	coreClient *stellarcore.Client
```

File: services/horizon/internal/test/integration/integration.go [hunk 14]
```diff
@@ -70,64 +99,92 @@ func NewTestForRemoteHorizon(t *testing.T, horizonURL string, passPhrase string,
...
+		}
+
+		i.waitForHorizon()
+	}
+
+	return i
+}
+
+// Runs a docker-compose command applied to the above configs
+func (i *Test) runComposeCommand(args ...string) {
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:157]: 💯

File: services/horizon/internal/test/integration/integration.go [hunk 15]
```diff
@@ -189,76 +243,93 @@ func (i *Test) startHorizon(
...
-		fmt.Sprintf("http://%s:%d", hostname, historyArchivePort),
-		"--ingest",
-		"--db-url",
-		horizonPostgresURL,
-		"--stellar-core-db-url",
-		fmt.Sprintf(
+	coreBinaryPath := i.coreConfig.binaryPath
+	captiveCoreConfigPath := i.coreConfig.configPath
+
+	var DEFAULT_ARGS = map[string]string{
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:262]:
```suggestion
	defaultArgs := map[string]string{
```

File: services/horizon/internal/test/integration/integration.go [hunk 16]
```diff
@@ -189,76 +243,93 @@ func (i *Test) startHorizon(
...
-		"21626",
-
-		"--enable-captive-core-ingestion=" + strconv.FormatBool(len(captiveCoreBinaryPath) > 0),
-
-		"--network-passphrase",
-		i.passPhrase,
-		"--apply-migrations",
-		"--admin-port",
-		strconv.Itoa(i.AdminPort()),
+	merged := mergeMaps(DEFAULT_ARGS, i.config.HorizonParameters)
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:286]:
```suggestion
	merged := mergeMaps(defaultArgs, i.config.HorizonParameters)
```

File: services/horizon/internal/test/integration/integration.go (hunk 2 above)
Comment (paulbellamy) [reply services/horizon/internal/test/integration/integration.go:310]: Eugh yeah. modifying the global os env here sucks. Would be good to stop doing that, but definitely out of scope for this PR.

File: services/horizon/internal/test/integration/integration.go [hunk 17]
```diff
@@ -302,6 +375,7 @@ func (i *Test) waitForCore() {
 			time.Sleep(time.Second)
 			continue
 		}
+		i.t.Log("Core is up.")
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:378]: 🆙 nice.

File: services/horizon/internal/test/integration/integration.go [hunk 18]
```diff
@@ -302,6 +375,7 @@ func (i *Test) waitForCore() {
...
 		return
 	}
 	i.t.Fatal("Core could not sync after several attempts")
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:381]:
```suggestion
	i.t.Fatal("Core could not sync after 30s")
```

File: services/horizon/internal/test/integration/integration.go [hunk 19]
```diff
@@ -660,3 +730,24 @@ func findDockerComposePath() string {
 	// Directly jump down to the folder that should contain the configs
 	return filepath.Join(current, "services", "horizon", "docker")
 }
+
+// mergeMaps returns a new map which contains the keys and values of both input
+// maps, preferring the latter newValues map if there are duplicate entries.
+func mergeMaps(defaults map[string]string, newValues map[string]string) map[string]string {
```
Comment (paulbellamy) [inline services/horizon/internal/test/integration/integration.go:736]: Ironically it's even simpler in the more generic case... 😂 

```go
func mergeMaps(ms ...map[string]string) map[string]string {
  merged := map[string]string{}
  for _, m := range ms {
    for k, v := range m {
      merged[k] = v
    }
  }
  return merged
}
```

File: services/horizon/internal/integration/parameters_test.go (hunk 8 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:87]: You're sorta there but not exactly, @paulbellamy.

The main problem is that Horizon dies via `log.Fatal` if the configuration options are invalid. Since the purpose of this framework is exactly to test whether or not Horizon treats invalid combinations as invalid (and that it treats valid combinations like it should), we basically need a way to say "did Horizon die as expected?" Since `log.Fatal` kills the process, the test also immediately dies. Thus, we need this (hacky, but functional) `suite.Exits()` approach to detect Horizon exiting _expectedly_.

File: services/horizon/internal/integration/parameters_test.go (hunk 5 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:115]: Oh good tip: `StartHorizon` and `waitForHorizon` are called separately, so I didn't see the latter.

File: services/horizon/internal/integration/parameters_test.go (hunk 6 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:76]: Yep, good points!

File: services/horizon/internal/integration/parameters_test.go (hunk 9 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:120]: I think it looks this way because there aren't enough example tests yet, but the intent of the framework here is essentially to associate certain "validation functions" with each parameter. You grouped the parameters together into `defaultCaptiveCoreConfig`, but in fact this is supposed to be the unique part. Otherwise, we aren't actually testing the parameter combinations!

However, I do think there's merit to having this default, and then instead modifying only what we need within the test. I'll tinker with it and push my ideas.

File: services/horizon/internal/test/integration/integration.go (hunk 10 above)
Comment (Shaptic) [reply services/horizon/internal/test/integration/integration.go:50]: yep, hated my naming; great suggestion

File: services/horizon/internal/integration/parameters_test.go (hunk 11 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:193]: great idea :+1:

File: services/horizon/internal/test/integration/integration.go (hunk 19 above)
Comment (Shaptic) [reply services/horizon/internal/test/integration/integration.go:736]: Clever! Love it

File: services/horizon/internal/integration/parameters_test.go [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+	// enabled, and once without. *These* tests only run with Captive Core
+	// configured properly (specifically, w/ the CAPTIVE_CORE_BIN envvar set).
+	if _, ok := os.LookupEnv("HORIZON_INTEGRATION_ENABLE_CAPTIVE_CORE"); !ok {
+		suite.T().Skip()
+	}
+
+	defer createCaptiveCoreConfig(
+		"./captive-core.toml", BUCKET_DIR_DISALLOWED_TOML)()
+
+	const STORAGE_PATH string = "./test_no-bucket-dir"
```
[the remark this answers is not in the document]
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:78]: :joy:

File: services/horizon/internal/test/integration/integration.go (hunk 2 above)
Comment (Shaptic) [reply services/horizon/internal/test/integration/integration.go]: wasn't a big change, so I introduced `EnvironmentManager` to help with this!

File: services/horizon/internal/test/integration/integration.go (hunk 3 above)
Comment (Shaptic) [reply services/horizon/internal/test/integration/integration.go]: you're right!

File: services/horizon/internal/integration/parameters_test.go (hunk 8 above)
Comment (paulbellamy) [reply services/horizon/internal/integration/parameters_test.go:83]: Yeah, I'm saying that, in an ideal world, only place in the codebase that should do `log.Fatal`, or `os.Exit` is `func main()`

File: services/horizon/internal/integration/parameters_test.go (hunk 9 above)
Comment (paulbellamy) [reply services/horizon/internal/integration/parameters_test.go:149]: Got it, ok. In that case, maybe we should use `t.Run("some name", ...)` to label them instead of the matrix function?

File: services/horizon/internal/integration/parameters_test.go (hunk 8 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:83]: Sure, agreed! But that's a bigger refactor than is in scope here, unfortunately. 

The alternative is to run the entirety of Horizon as a subprocess rather than "natively" in a `go func() {}`; that refactor may actually be "better" in a way because it enables us to test subcommands, too (e.g. `migrate`).

File: services/horizon/internal/integration/parameters_test.go (hunk 9 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:146]: Hm, seems reasonable. Done in 2d98e427.

Comment (paulbellamy) [reply services/horizon/internal/integration/parameters_test.go:146]: Cool. Looks better at least. As a nit, I'd probably inline `validateNoBucketDirPath`, and the other one (since they're only used in one place). But it's good enough for now.

File: services/horizon/internal/integration/parameters_test.go (hunk 8 above)
Comment (paulbellamy) [reply services/horizon/internal/integration/parameters_test.go:82]: Nit, should this bit about log.Fatal and it dying be a comment on the `Exits` method?

**Review by paulbellamy — APPROVED** [review]
Definitely more refactoring we could do to make testing this simpler, but much better than we had, and more than good enough for now. 💯

File: services/horizon/internal/integration/parameters_test.go (hunk 9 above)
Comment (Shaptic) [reply services/horizon/internal/integration/parameters_test.go:146]: Fair, but again the idea is that these would be reused in new tests that also wanted to confirm the state of e.g. `--captive-core-storage-path`.
